behave/runner.py

```diff
@@ -918,11 +918,17 @@ def load_hooks(self, filename=None):
     def load_step_definitions(self, extra_step_paths=None):
         if extra_step_paths is None:
             extra_step_paths = []
+
         # -- Allow steps to import other stuff from the steps dir
         # NOTE: Default matcher can be overridden in "environment.py" hook.
+        # Load step modules from subdirectories within the steps directory
         steps_dir = os.path.join(self.base_dir, self.config.steps_dir)
-        step_paths = [steps_dir] + list(extra_step_paths)
-        load_step_modules(step_paths)
+        for root, dirs, files in os.walk(steps_dir):
+            for file in files:
+                if file.endswith('.py') and not file.startswith('__init__'):
+                    step_file = os.path.join(root, file)
+                    if os.path.isfile(step_file):  # Ensure it's a file
+                        load_step_modules([root])  # Load from the directory
```

Comment:
The nested for-loop that calls [load_step_modules](https://github.com/behave/behave/blob/main/behave/runner_util.py#L566) is much too complex. It's incredibly hard to understand what it does.

Can't you collect the list of `root` folders and submit that list to `load_step_modules` in a single call?

Reply:
@bittner 
I’ve addressed the feedback by simplifying the logic. 
Now, I collect all root directories containing step files and pass them to load_step_modules in a single call.